OptimizelySDK/Bucketing/DecisionService.cs

```diff
@@ -253,5 +246,156 @@ public void SaveVariation(Experiment experiment, Variation variation, UserProfil
             }
         }
+
+        /// <summary>
+        /// Try to bucket the user into a rollout rule.
+        /// Evaluate the user for rules in priority order by seeing if the user satisfies the audience.
+        /// Fall back onto the everyone else rule if the user is ever excluded from a rule due to traffic allocation.
+        /// </summary>
+        /// <param name = "featureFlag" >The feature flag the user wants to access.</param>
+        /// <param name = "userId" >User Identifier</param>
+        /// <param name = "filteredAttributes" >The user's attributes. This should be filtered to just attributes in the Datafile.</param>
+        /// <returns>null if the user is not bucketed into the rollout or if the feature flag was not attached to a rollout.
+        /// otherwise the Variation the user is bucketed into</returns>
+        public virtual Variation GetVariationForFeatureRollout(FeatureFlag featureFlag, string userId, UserAttributes filteredAttributes)
+        {
+            if (string.IsNullOrEmpty(featureFlag.RolloutId))
+            {
+                Logger.Log(LogLevel.INFO, $"The feature flag \"{featureFlag.Key}\" is not used in a rollout.");
+                return null;
+            }
+
+            Rollout rollout = ProjectConfig.GetRolloutFromId(featureFlag.RolloutId);
+
+            if (string.IsNullOrEmpty(rollout.Id))
+            {
+                Logger.Log(LogLevel.ERROR, $"The rollout with id \"{featureFlag.RolloutId}\" is not found in the datafile for feature flag \"{featureFlag.Key}\"");
+                return null;
+            }
+
+            Variation variation = null;
+            var rolloutRulesLength = rollout.Experiments.Count;
+            
+            // Get Bucketing ID from user attributes.
+            string bucketingId = GetBucketingId(userId, filteredAttributes);
+
+            // For all rules before the everyone else rule
+            for (int i=0; i < rolloutRulesLength - 1; i++)
+            {
+                var rolloutRule = rollout.Experiments[i];
+
+                if (ExperimentUtils.IsUserInExperiment(ProjectConfig, rolloutRule, filteredAttributes))
+                {
+                    Logger.Log(LogLevel.DEBUG, $"Attempting to bucket user \"{userId}\" into rollout rule \"{rolloutRule.Key}\".");
+                    variation = Bucketer.Bucket(ProjectConfig, rolloutRule, bucketingId, userId);
+
+                    if (variation == null)
+                    {
+                        Logger.Log(LogLevel.DEBUG, $"User \"{userId}\" is excluded due to traffic allocation. Checking \"Eveyrone Else\" rule now.");
+                        break;
+                    }
+
+                    return variation;
+                }
+                else
+                {
+                    Logger.Log(LogLevel.DEBUG, $"User \"{userId}\" does not meet the audience conditions to be in rollout rule \"{rolloutRule.Key}\".");
```

Comment:
rolloutRule.key doesn't make sense since it will just be the same as the ID. saying they are excluded due to audience conditions for the audience  name would be more easily debugged

Reply:
Got it. Replaced rolloutRule.Key with audience name.